File: backend/app/services/integrations/slack_integration.py
from typing import Dict, List, Any, Optional
from datetime import datetime
import json

from app.services.base_integration import BaseIntegration, IntegrationStatus
from app.models.api_keys import SupportedService
from app.core.logging import get_logger

logger = get_logger(__name__)

class SlackIntegration(BaseIntegration):
    """Slack API integration for messaging, channel management, and team communication automation"""
    
    def __init__(self):
        super().__init__(SupportedService.SLACK)
    
    @property
    def base_url(self) -> str:
        return "https://slack.com/api"
    
    @property
    def required_scopes(self) -> List[str]:
        return [
            "channels:read",
            "channels:write", 
            "chat:write",
            "users:read",
            "users:read.email",
            "files:write"
        ]
    
    async def _prepare_auth_headers(self, api_key: str) -> Dict[str, str]:
        """Prepare Slack API headers"""
        return {
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json"
        }
    
    async def _test_api_connection(self, api_key: str) -> Dict[str, Any]:
        """Test Slack API connection by getting auth info"""
        try:
            result = await self.make_api_request(
                method="GET",
                endpoint="auth.test",
                api_key=api_key
            )
            
            if result.get("success"):
                auth_data = result.get("data", {})
                if auth_data.get("ok"):
                    return {
                        "connected": True,
                        "status": "success",
                        "message": "Successfully connected to Slack",
                        "workspace_info": {
                            "team": auth_data.get("team", "Unknown"),
                            "user": auth_data.get("user", "Unknown"),
                            "user_id": auth_data.get("user_id", "Unknown"),
                            "team_id": auth_data.get("team_id", "Unknown")
                        }
                    }
                else:
                    return {
                        "connected": False,
                        "status": "error",
                        "message": auth_data.get("error", "Authentication failed")
                    }
            else:
                return {
                    "connected": False,
                    "status": "error",
                    "message": result.get("message", "Failed to connect to Slack")
                }
                
        except Exception as e:
            logger.error(f"Slack connection test failed: {str(e)}")
            return {
                "connected": False,
                "status": "error", 
                "message": str(e)
            }
    
    # Core Slack Operations
    
    async def get_channels(self, session_id: str, types: str = "public_channel,private_channel") -> Dict[str, Any]:
        """Get all channels in the workspace"""
        api_key = await self.get_api_key(session_id)
        if not api_key:
            return {"success": False, "error": "no_api_key"}
        
        try:
            result = await self.make_api_request(
                method="GET",
                endpoint="conversations.list",
                api_key=api_key,
                params={"types": types, "limit": 100}
            )
            
            if result.get("success"):
                data = result.get("data", {})
                if data.get("ok"):
                    channels = []
                    for channel in data.get("channels", []):
                        channels.append({
                            "id": channel.get("id"),
                            "name": channel.get("name"),
                            "is_private": channel.get("is_private", False),
                            "is_member": channel.get("is_member", False),
                            "topic": channel.get("topic", {}).get("value", ""),
                            "purpose": channel.get("purpose", {}).get("value", ""),
                            "member_count": channel.get("num_members", 0)
                        })
                    
                    return {
                        "success": True,
                        "channels": channels,
                        "count": len(channels)
                    }
                else:
                    return {"success": False, "error": data.get("error", "Failed to get channels")}
            else:
                return result
                
        except Exception as e:
            logger.error(f"Error getting Slack channels: {str(e)}")
            return {"success": False, "error": str(e)}
    
    async def get_users(self, session_id: str) -> Dict[str, Any]:
        """Get all users in the workspace"""
        api_key = await self.get_api_key(session_id)
        if not api_key:
            return {"success": False, "error": "no_api_key"}
        
        try:
            result = await self.make_api_request(
                method="GET",
                endpoint="users.list",
                api_key=api_key,
                params={"limit": 100}
            )
            
            if result.get("success"):
                data = result.get("data", {})
                if data.get("ok"):
                    users = []
                    for user in data.get("members", []):
                        if not user.get("deleted", False) and not user.get("is_bot", False):
                            users.append({
                                "id": user.get("id"),
                                "name": user.get("name"),
                                "real_name": user.get("real_name", ""),
                                "display_name": user.get("profile", {}).get("display_name", ""),
                                "email": user.get("profile", {}).get("email", ""),
                                "is_admin": user.get("is_admin", False),
                                "is_owner": user.get("is_owner", False)
                            })
                    
                    return {
                        "success": True,
                        "users": users,
                        "count": len(users)
                    }
                else:
                    return {"success": False, "error": data.get("error", "Failed to get users")}
            else:
                return result
                
        except Exception as e:
            logger.error(f"Error getting Slack users: {str(e)}")
            return {"success": False, "error": str(e)}
    
    async def send_message(
        self, 
        session_id: str, 
        channel: str, 
        text: str,
        thread_ts: Optional[str] = None,
        blocks: Optional[List[Dict]] = None
    ) -> Dict[str, Any]:
        """Send a message to a channel or user"""
        api_key = await self.get_api_key(session_id)
        if not api_key:
            return {"success": False, "error": "no_api_key"}
        
        try:
            data = {
                "channel": channel,
                "text": text
            }
            
            if thread_ts:
                data["thread_ts"] = thread_ts
            
            if blocks:
                data["blocks"] = blocks
            
            result = await self.make_api_request(
                method="POST",
                endpoint="chat.postMessage",
                api_key=api_key,
                data=data
            )
            
            if result.get("success"):
                response_data = result.get("data", {})
                if response_data.get("ok"):
                    return {
                        "success": True,
                        "message": {
                            "ts": response_data.get("ts"),
                            "channel": response_data.get("channel"),
                            "text": text
                        }
                    }
                else:
                    return {"success": False, "error": response_data.get("error", "Failed to send message")}
            else:
                return result
                
        except Exception as e:
            logger.error(f"Error sending Slack message: {str(e)}")
            return {"success": False, "error": str(e)}
    
    async def create_channel(
        self, 
        session_id: str, 
        name: str, 
        is_private: bool = False,
        topic: Optional[str] = None,
        purpose: Optional[str] = None
    ) -> Dict[str, Any]:
        """Create a new channel"""
        api_key = await self.get_api_key(session_id)
        if not api_key:
            return {"success": False, "error": "no_api_key"}
        
        try:
            data = {
                "name": name,
                "is_private": is_private
            }
            
            result = await self.make_api_request(
                method="POST",
                endpoint="conversations.create",
                api_key=api_key,
                data=data
            )
            
            if result.get("success"):
                response_data = result.get("data", {})
                if response_data.get("ok"):
                    channel_data = response_data.get("channel", {})
                    channel_id = channel_data.get("id")
                    
                    # Set topic and purpose if provided
                    if topic and channel_id:
                        await self.make_api_request(
                            method="POST",
                            endpoint="conversations.setTopic",
                            api_key=api_key,
                            data={"channel": channel_id, "topic": topic}
                        )
                    
                    if purpose and channel_id:
                        await self.make_api_request(
                            method="POST",
                            endpoint="conversations.setPurpose",
                            api_key=api_key,
                            data={"channel": channel_id, "purpose": purpose}
                        )
                    
                    return {
                        "success": True,
                        "channel": {
                            "id": channel_id,
                            "name": channel_data.get("name"),
                            "is_private": channel_data.get("is_private", False),
                            "created": channel_data.get("created")
                        }
                    }
                else:
                    return {"success": False, "error": response_data.get("error", "Failed to create channel")}
            else:
                return result
                
        except Exception as e:
            logger.error(f"Error creating Slack channel: {str(e)}")
            return {"success": False, "error": str(e)}
    
    async def invite_to_channel(
        self, 
        session_id: str, 
        channel: str, 
        users: List[str]
    ) -> Dict[str, Any]:
        """Invite users to a channel"""
        api_key = await self.get_api_key(session_id)
        if not api_key:
            return {"success": False, "error": "no_api_key"}
        
        try:
            data = {
                "channel": channel,
                "users": ",".join(users)
            }
            
            result = await self.make_api_request(
                method="POST",
                endpoint="conversations.invite",
                api_key=api_key,
                data=data
            )
            
            if result.get("success"):
                response_data = result.get("data", {})
                if response_data.get("ok"):
                    return {
                        "success": True,
                        "message": f"Successfully invited {len(users)} users to channel"
                    }
                else:
                    return {"success": False, "error": response_data.get("error", "Failed to invite users")}
            else:
                return result
                
        except Exception as e:
            logger.error(f"Error inviting users to Slack channel: {str(e)}")
            return {"success": False, "error": str(e)}
    
    async def schedule_message(
        self, 
        session_id: str, 
        channel: str, 
        text: str,
        post_at: int,  # Unix timestamp
        blocks: Optional[List[Dict]] = None
    ) -> Dict[str, Any]:
        """Schedule a message to be sent later"""
        api_key = await self.get_api_key(session_id)
        if not api_key:
            return {"success": False, "error": "no_api_key"}
        
        try:
            data = {
                "channel": channel,
                "text": text,
                "post_at": post_at
            }
            
            if blocks:
                data["blocks"] = blocks
            
            result = await self.make_api_request(
                method="POST",
                endpoint="chat.scheduleMessage",
                api_key=api_key,
                data=data
            )
            
            if result.get("success"):
                response_data = result.get("data", {})
                if response_data.get("ok"):
                    return {
                        "success": True,
                        "scheduled_message": {
                            "id": response_data.get("scheduled_message_id"),
                            "channel": channel,
                            "post_at": post_at
                        }
                    }
                else:
                    return {"success": False, "error": response_data.get("error", "Failed to schedule message")}
            else:
                return result
                
        except Exception as e:
            logger.error(f"Error scheduling Slack message: {str(e)}")
            return {"success": False, "error": str(e)}
    
    # Trigger.dev Workflow Integration
    
    async def _create_service_workflow(
        self, 
        workflow_payload: Dict[str, Any], 
        api_key: str
    ) -> Dict[str, Any]:
        """Create Slack-specific Trigger.dev workflows"""
        
        workflow_name = workflow_payload.get("name")
        config = workflow_payload.get("config", {})
        
        # Define available Slack workflows
        slack_workflows = {
            "team_notifications": {
                "description": "Send automated notifications to team channels",
                "triggers": ["milestone_reached", "deadline_approaching", "error_occurred"],
                "actions": ["send_message", "create_channel", "schedule_message"]
            },
            "project_updates": {
                "description": "Share project progress and updates with team",
                "triggers": ["status_changed", "task_completed", "manual"],
                "actions": ["send_message", "create_thread", "mention_users"]
            },
            "meeting_coordination": {
                "description": "Coordinate meetings and send reminders",
                "triggers": ["meeting_scheduled", "reminder_time"],
                "actions": ["send_message", "schedule_message", "create_channel"]
            },
            "alert_system": {
                "description": "Send alerts and urgent notifications",
                "triggers": ["error_detected", "threshold_exceeded", "manual"],
                "actions": ["send_message", "mention_channel", "escalate"]
            }
        }
        
        workflow_type = config.get("workflow_type", "team_notifications")
        
        if workflow_type not in slack_workflows:
            return {
                "success": False,
                "error": "invalid_workflow_type",
                "available_workflows": list(slack_workflows.keys())
            }
        
        # Create workflow configuration
        workflow_config = {
            "name": workflow_name,
            "service": "slack",
            "type": workflow_type,
            "config": slack_workflows[workflow_type],
            "user_config": config,
            "created_at": datetime.utcnow().isoformat()
        }
        
        # In a real implementation, this would call Trigger.dev API
        return {
            "success": True,
            "workflow": workflow_config,
            "message": f"Slack workflow '{workflow_name}' created successfully"
        }
    
    # Helper Methods
    
    def create_rich_message_blocks(
        self, 
        title: str, 
        content: str, 
        color: str = "good",
        fields: Optional[List[Dict]] = None
    ) -> List[Dict[str, Any]]:
        """Create rich message blocks for better formatting"""
        
        blocks = [
            {
                "type": "section",
                "text": {
                    "type": "mrkdwn",
                    "text": f"*{title}*\n{content}"
                }
            }
        ]
        
        if fields:
            blocks.append({
                "type": "section",
                "fields": [
                    {
                        "type": "mrkdwn",
                        "text": f"*{field['title']}*\n{field['value']}"
                    }
                    for field in fields
                ]
            })
        
        return blocks
    
    def create_action_blocks(self, actions: List[Dict[str, str]]) -> Dict[str, Any]:
        """Create action buttons for interactive messages"""
        
        elements = []
        for action in actions:
            elements.append({
                "type": "button",
                "text": {
                    "type": "plain_text",
                    "text": action.get("text", "Action")
                },
                "value": action.get("value", ""),
                "action_id": action.get("action_id", f"action_{len(elements)}")
            })
        
        return {
            "type": "actions",
            "elements": elements
        }

# Global instance
slack_integration = SlackIntegration() 

File: trigger-jobs/src/trigger/slack-integration.ts
import { task } from "@trigger.dev/sdk/v3";
import { WebClient } from "@slack/web-api";

// Slack integration task for Agent OS
export const sendSlackMessage = task({
  id: "agent-os-slack-message",
  run: async (payload: {
    channel: string;
    message: string;
    agent_name?: string;
    user_id?: string;
  }) => {
    // Initialize Slack client with bot token
    const slack = new WebClient(process.env.SLACK_BOT_TOKEN);
    
    // Format message with Agent OS branding
    const formattedMessage = payload.agent_name 
      ? `🤖 *${payload.agent_name}* from Agent OS:\n\n${payload.message}`
      : `🤖 *Agent OS*:\n\n${payload.message}`;
    
    try {
      // Send message to Slack
      const result = await slack.chat.postMessage({
        channel: payload.channel,
        text: formattedMessage,
        blocks: [
          {
            type: "section",
            text: {
              type: "mrkdwn",
              text: formattedMessage
            }
          },
          {
            type: "context",
            elements: [
              {
                type: "mrkdwn",
                text: `Sent by Agent OS • ${new Date().toLocaleString()}`
              }
            ]
          }
        ]
      });
      
      return {
        success: true,
        message_ts: result.ts,
        channel: result.channel,
        agent_name: payload.agent_name,
        timestamp: new Date().toISOString()
      };
      
    } catch (error) {
      console.error("Slack message failed:", error);
      throw error;
    }
  },
});

// Task to create Slack channels for Agent OS
export const createSlackChannel = task({
  id: "agent-os-create-channel",
  run: async (payload: {
    channel_name: string;
    purpose?: string;
    is_private?: boolean;
  }) => {
    const slack = new WebClient(process.env.SLACK_BOT_TOKEN);
    
    try {
      // Create the channel
      const result = await slack.conversations.create({
        name: payload.channel_name,
        is_private: payload.is_private || false,
        purpose: payload.purpose || `Channel created by Agent OS`
      });
      
      // Send welcome message
      if (result.channel?.id) {
        await slack.chat.postMessage({
          channel: result.channel.id,
          text: `🎉 Welcome to #${payload.channel_name}!\n\nThis channel was created by Agent OS to help organize your team's workflow.`,
          blocks: [
            {
              type: "section",
              text: {
                type: "mrkdwn",
                text: `🎉 *Welcome to #${payload.channel_name}!*\n\nThis channel was created by Agent OS to help organize your team's workflow.`
              }
            },
            {
              type: "divider"
            },
            {
              type: "section",
              text: {
                type: "mrkdwn",
                text: "Agent OS can help you:\n• Send automated notifications\n• Share project updates\n• Coordinate team activities\n• Track important milestones"
              }
            }
          ]
        });
      }
      
      return {
        success: true,
        channel_id: result.channel?.id,
        channel_name: payload.channel_name,
        created_at: new Date().toISOString()
      };
      
    } catch (error) {
      console.error("Channel creation failed:", error);
      throw error;
    }
  },
});

// Task to get Slack workspace info for Agent OS setup
export const getSlackWorkspaceInfo = task({
  id: "agent-os-slack-workspace-info",
  run: async () => {
    const slack = new WebClient(process.env.SLACK_BOT_TOKEN);
    
    try {
      // Get workspace info
      const teamInfo = await slack.team.info();
      
      // Get channels list
      const channels = await slack.conversations.list({
        types: "public_channel,private_channel",
        limit: 100
      });
      
      // Get bot info
      const authTest = await slack.auth.test();
      
      return {
        success: true,
        workspace: {
          name: teamInfo.team?.name,
          id: teamInfo.team?.id,
          domain: teamInfo.team?.domain
        },
        bot: {
          user_id: authTest.user_id,
          bot_id: authTest.bot_id,
          user: authTest.user
        },
        channels: channels.channels?.map(channel => ({
          id: channel.id,
          name: channel.name,
          is_private: channel.is_private,
          is_member: channel.is_member
        })) || [],
        integration_status: "connected",
        timestamp: new Date().toISOString()
      };
      
    } catch (error) {
      console.error("Workspace info failed:", error);
      throw error;
    }
  },
});

// Demo task to show Agent OS Slack integration working
export const demoAgentOSSlackIntegration = task({
  id: "demo-agent-os-slack-integration",
  run: async (payload: {
    demo_channel?: string;
    user_name?: string;
  }) => {
    const slack = new WebClient(process.env.SLACK_BOT_TOKEN);
    const channel = payload.demo_channel || "general";
    const userName = payload.user_name || "Demo User";
    
    try {
      // Send demo message showing Agent OS capabilities
      const result = await slack.chat.postMessage({
        channel: channel,
        text: `🚀 Agent OS is now connected to your Slack workspace!`,
        blocks: [
          {
            type: "header",
            text: {
              type: "plain_text",
              text: "🚀 Agent OS Connected!"
            }
          },
          {
            type: "section",
            text: {
              type: "mrkdwn",
              text: `Hey ${userName}! Agent OS is now successfully integrated with your Slack workspace.`
            }
          },
          {
            type: "divider"
          },
          {
            type: "section",
            text: {
              type: "mrkdwn",
              text: "*What can Agent OS do in Slack?*\n\n🤖 *Alex* - Strategic planning updates\n🎨 *Dana* - Creative content sharing\n📊 *Riley* - Data insights and reports\n⚙️ *Jamie* - Operations notifications"
            }
          },
          {
            type: "section",
            text: {
              type: "mrkdwn",
              text: "*Ready to get started?*\nYour agents can now send updates, share files, and keep your team in sync automatically!"
            }
          },
          {
            type: "context",
            elements: [
              {
                type: "mrkdwn",
                text: `Connected via Agent OS • ${new Date().toLocaleString()}`
              }
            ]
          }
        ]
      });
      
      return {
        success: true,
        demo_completed: true,
        message_sent: true,
        channel: channel,
        message_ts: result.ts,
        next_steps: [
          "Your Slack workspace is connected to Agent OS",
          "All 4 agents (Alex, Dana, Riley, Jamie) can now send messages",
          "You can test the integration from the Agent OS chat interface",
          "Set up automated workflows in the Agent OS dashboard"
        ],
        timestamp: new Date().toISOString()
      };
      
    } catch (error) {
      console.error("Demo integration failed:", error);
      throw error;
    }
  },
}); 